### PR TITLE
Add permissions and type to jwt token generation

### PR DESCRIPTION
**Pull request by ranbreuer**
Sending this as a merge to master - although it should have been done before merged to devBranch. :)

- Added mandatory "embed" type to jwt
- Added ability for datasetId & scopes claims
- Test coverage expanded

Note: This wont be merged to master until relevant features are in production.

### CODE REVIEW

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 1]
```diff
@@ -101,21 +158,27 @@ public static PowerBIToken CreateReportEmbedToken(string workspaceCollectionName
...
-            Guard.ValidateString(reportId, "reportId");
 
             if (expiration < DateTime.UtcNow)
             {
                 throw new ArgumentException("Expiration must be a date/time in the future", nameof(expiration));
             }
 
+            if (string.IsNullOrWhiteSpace(reportId) && string.IsNullOrWhiteSpace(datasetId))
+            {
+                throw new ArgumentException("Either ReportId or DatasetId must be a date/time in the future", nameof(reportId) + "\\ " + nameof(datasetId));
```
Comment (noanu) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: must be "a date/time"...?

Comment (ranbreuer) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: must be set..

good catch!

File: sdk/PowerBI.Core/Properties/AssemblyInfo.cs [hunk 2]
```diff
@@ -25,6 +25,6 @@
 // You can specify all the values or you can default the Build and Revision Numbers 
 // by using the '*' as shown below:
 // [assembly: AssemblyVersion("1.0.*")]
-[assembly: AssemblyVersion("1.0.0.0")]
-[assembly: AssemblyFileVersion("1.0.0.0")]
+[assembly: AssemblyVersion("1.1.0.0")]
+[assembly: AssemblyFileVersion("1.1.0.0")]
```
Comment (ali-hamud) [inline sdk/PowerBI.Core/Properties/AssemblyInfo.cs]: version should be 2.0.0.0 for a breaking change.
please look at here: http://semver.org/
or here:
https://blogs.msdn.microsoft.com/jjameson/2009/04/03/best-practices-for-net-assembly-versioning/

Comment (ranbreuer) [inline sdk/PowerBI.Core/Properties/AssemblyInfo.cs]: good catch! thanks!

File: sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs [hunk 3]
```diff
@@ -46,6 +49,41 @@ public void CanCreateReportEmbedToken()
...
+            var widClaim = decodedToken.Claims.FirstOrDefault(c => c.Type == PowerBIToken.ClaimTypes.WorkspaceId);
+            var didCliam = decodedToken.Claims.FirstOrDefault(c => c.Type == PowerBIToken.ClaimTypes.DatasetId);
+            var usernameClaim = decodedToken.Claims.FirstOrDefault(c => c.Type == PowerBIToken.ClaimTypes.Username);
+            var rolesClaim = decodedToken.Claims.FirstOrDefault(c => c.Type == PowerBIToken.ClaimTypes.Roles);
+
+            Assert.AreEqual("PowerBISDK", decodedToken.Issuer);
+            Assert.IsTrue(decodedToken.Audiences.Contains("https://analysis.windows.net/powerbi/api"));
+            Assert.IsTrue(decodedToken.ValidTo >= DateTime.UtcNow);
+            Assert.IsTrue(decodedToken.ValidTo <= DateTime.UtcNow.AddHours(1));
+            Assert.AreEqual("0.2.0", versionClaim.Value);
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs]: Consider adding const variable for the version number and use it as the expected value throughout the different tests

File: sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs [hunk 4]
```diff
@@ -126,13 +164,69 @@ public void CanCreateReportEmbedTokenWithRlsWithMultipleRoles()
         }
 
         [TestMethod]
+        public void CanCreateReportEmbedTokenWithScopes()
+        {
+            var workspaceId = Guid.NewGuid().ToString();
+            var reportId = Guid.NewGuid().ToString();
+            var scopes = Scopes.DatasetRead;
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs]: If you're using the dataset read scope please change the resource to dataset id or modify the scope for the default read report one

File: sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs [hunk 5]
```diff
@@ -126,13 +164,69 @@ public void CanCreateReportEmbedTokenWithRlsWithMultipleRoles()
...
+
+            var token = PowerBIToken.CreateReportEmbedToken("Contoso", workspaceId, reportId, scopes: scopes);
+
+            Assert.IsNotNull(token);
+            var jwt = token.Generate(this.accessKey);
+            Assert.IsFalse(string.IsNullOrEmpty(jwt));
+
+            var decodedToken = new JwtSecurityToken(jwt);
+
+            var scopesClaim = decodedToken.Claims.FirstOrDefault(c => c.Type == PowerBIToken.ClaimTypes.Scopes).Value.Split(' ').ToList();
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core.Tests/Security/PowerBITokenTests.cs]: consider adding utility function for extracting scopes .... Value.Split(' ')

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 6]
```diff
@@ -70,29 +87,69 @@ public PowerBIToken()
         /// <param name="workspaceCollectionName">The workspace collection name</param>
         /// <param name="workspaceId">The workspace id</param>
         /// <param name="reportId">The report id</param>
+        /// <param name="datasetId">The dataset id</param>
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: Seems redundant

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 7]
```diff
@@ -70,29 +87,69 @@ public PowerBIToken()
...
         /// <param name="username">The RLS username</param>
         /// <param name="roles">The RLS roles</param>
+        /// <param name="scopes">The permission scopes</param>
         /// <returns>The Power BI access token</returns>
-        public static PowerBIToken CreateReportEmbedToken(string workspaceCollectionName, string workspaceId, string reportId, string username = null, IEnumerable<string> roles = null)
+        public static PowerBIToken CreateReportEmbedToken(string workspaceCollectionName, string workspaceId, string reportId = null, string datasetId = null, string username = null, IEnumerable<string> roles = null, string scopes = null)
         {
             var expires = DateTime.UtcNow.Add(TimeSpan.FromSeconds(DefaultExpirationSeconds));
-            return CreateReportEmbedToken(workspaceCollectionName, workspaceId, reportId, expires, username, roles);
+            return CreateReportEmbedToken(workspaceCollectionName, workspaceId, reportId: reportId, datasetId: string.Empty, expiration: expires, username: username, roles: roles, scopes: scopes);
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: Why do we define datasetId as a parameter if it's being ignored later? Please change the signature accordingly

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 8]
```diff
@@ -70,29 +87,69 @@ public PowerBIToken()
...
         /// </summary>
         /// <param name="workspaceCollectionName">The workspace collection name</param>
         /// <param name="workspaceId">The workspace id</param>
         /// <param name="reportId">The report id</param>
+        /// <param name="datasetId">The dataset id</param>
+        /// <param name="username">The RLS username</param>
+        /// <param name="roles">The RLS roles</param>
+        /// <param name="scopes">The permission scopes</param>
+        /// <returns>The Power BI access token</returns>
+        public static PowerBIToken CreateReportEmbedTokenWithDataset(string workspaceCollectionName, string workspaceId, string datasetId, string reportId = null, string username = null, IEnumerable<string> roles = null, string scopes = null)
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: Isn't it a bit confusing? When will the developer pass both report and dataset identifiers?

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 9]
```diff
@@ -128,7 +191,18 @@ public static PowerBIToken CreateReportEmbedToken(string workspaceCollectionName
...
+            }
+
+            if (!string.IsNullOrWhiteSpace(datasetId))
+            {
+                token.DatasetId = datasetId;
+                token.Claims.Add(new Claim(ClaimTypes.DatasetId, datasetId));
+            }
 
             // RLS claims: requires username and roles are optional
             if (!string.IsNullOrEmpty(username))
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: Why aren't we populating the user and roles claims here?

File: sdk/PowerBI.Core/Security/PowerBIToken.cs [hunk 10]
```diff
@@ -72,11 +89,64 @@ public PowerBIToken()
...
+        /// <param name="scopes">The permission scopes</param>
+        /// <returns>The Power BI access token</returns>
+        public static PowerBIToken CreateReportEmbedTokenForCreation(string workspaceCollectionName, string workspaceId, string datasetId, string username = null, IEnumerable<string> roles = null, string scopes = null)
+        {
+            var expires = DateTime.UtcNow.Add(TimeSpan.FromSeconds(DefaultExpirationSeconds));
+            return CreateReportEmbedToken(workspaceCollectionName, workspaceId, null, datasetId, expires, username, roles, scopes);
+        }
+
+        /// <summary>
+        /// Creates a embed token with default expiration used to embed Power BI components into your own applications
```
Comment (ArinaHantsis) [inline sdk/PowerBI.Core/Security/PowerBIToken.cs]: The above method uses default expiration, this one uses something else

File: sdk/swagger.json [hunk 11]
```diff
@@ -1187,6 +1306,48 @@
         "workspaceCollectionName": {
           "type": "string",
           "description": "The workspace collection name"
+        },
+        "displayName": {
```
Comment (ArinaHantsis) [inline sdk/swagger.json]: Didn't Mahir already checked it in to master? Why does it show-up within the diff?

**Pull request merged**